From 904e7208620e904f4ad8157619ec3bb285a66189 Mon Sep 17 00:00:00 2001
From: Qais Yousef <qyousef@google.com>
Date: Mon, 11 May 2026 13:59:07 +0000
Subject: [PATCH 141/223] ANDROID: sched: Add vendor hook to force update pelt
 signal

Blocking the update based on period can result in unwanted variations in
util signal.

Bug: 512050420
Bug: 512490607
Signed-off-by: Qais Yousef <qyousef@google.com>
Cherrypick-From: https://android-review.googlesource.com/q/commit:0e2699e020d5a38d5d49c89fbb2dbdde30eb530f
Merged-In: I45de54e04b4d5bfa1edcf1054bc0cf4db6dfab43
Change-Id: I45de54e04b4d5bfa1edcf1054bc0cf4db6dfab43
---
 include/trace/hooks/sched.h | 4 ++++
 kernel/sched/pelt.c         | 5 ++++-
 kernel/sched/vendor_hooks.c | 1 +
 3 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/include/trace/hooks/sched.h b/include/trace/hooks/sched.h
index b91156c0dd24..426f0d723f4e 100644
--- a/include/trace/hooks/sched.h
+++ b/include/trace/hooks/sched.h
@@ -487,6 +487,10 @@ DECLARE_RESTRICTED_HOOK(android_rvh_update_rq_clock_pelt,
 	TP_PROTO(struct rq *rq, s64 delta, int *ret),
 	TP_ARGS(rq, delta, ret), 1);
 
+DECLARE_RESTRICTED_HOOK(android_rvh_update_load_sum,
+	TP_PROTO(int *force_update),
+	TP_ARGS(force_update), 1);
+
 DECLARE_RESTRICTED_HOOK(android_rvh_update_load_avg_blocked_se,
 	TP_PROTO(u64 now, struct sched_entity *se, int *ret),
 	TP_ARGS(now, se, ret), 1);
diff --git a/kernel/sched/pelt.c b/kernel/sched/pelt.c
index ab2e7e7a7240..3cf78dc9e8a2 100644
--- a/kernel/sched/pelt.c
+++ b/kernel/sched/pelt.c
@@ -180,6 +180,7 @@ __always_inline int
 ___update_load_sum(u64 now, struct sched_avg *sa,
 		  unsigned long load, unsigned long runnable, int running)
 {
+	int force_update = 0;
 	u64 delta;
 
 	delta = now - sa->last_update_time;
@@ -216,6 +217,8 @@ ___update_load_sum(u64 now, struct sched_avg *sa,
 	if (!load)
 		runnable = running = 0;
 
+	trace_android_rvh_update_load_sum(&force_update);
+
 	/*
 	 * Now we know we crossed measurement unit boundaries. The *_avg
 	 * accrues by two steps:
@@ -224,7 +227,7 @@ ___update_load_sum(u64 now, struct sched_avg *sa,
 	 * crossed period boundaries, finish.
 	 */
 	if (!accumulate_sum(delta, sa, load, runnable, running))
-		return 0;
+		return force_update;
 
 	return 1;
 }
diff --git a/kernel/sched/vendor_hooks.c b/kernel/sched/vendor_hooks.c
index 5532724082a1..8987750de261 100644
--- a/kernel/sched/vendor_hooks.c
+++ b/kernel/sched/vendor_hooks.c
@@ -125,6 +125,7 @@ EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_scx_task_can_run_on);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_scx_exit_on_abnormal);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_switching_to_scx);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_rvh_update_rq_clock_pelt);
+EXPORT_TRACEPOINT_SYMBOL_GPL(android_rvh_update_load_sum);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_rvh_update_load_avg_blocked_se);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_rvh_update_load_avg_se);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_rvh_update_load_avg_cfs_rq);
-- 
2.50.1

